### ISSUE TITLE
Sensor Battery Voltage is unknown / Solar Generation negative reading / Energy import & export entities

### ISSUE THREAD

**Issue by isngofoz** [izevaka/powersensor-home-assistant#6]
Finally managed to configure the plug with 2 sensors, which at this stage appears to be showing normal data, Total Home and Solar Generation, will know more tomorrow during the day. However the battery voltages for the 2 sensors are unknown but plug_name_voltage is showing 241.07 V. 

Any ideas?

**Comment by izevaka** [comment]
I think that’s expected. Only the plug knows the voltage because it’s
plugged into the mains. The battery voltages are only good for estimating
battery life.

What’s your set up? One plug and two sensors?

**Comment by isngofoz** [comment]
Thanks for your reply. Yes, one plug two sensors as shown in the photo below. I did note that when I plugged a power bank to charge the solar sensor, the battery voltage (3.75V) showed up. I will check again when it is fully charged. to see if it was unknown because the voltage was too low to be read.

![20230406_233454000_iOS](https://user-images.githubusercontent.com/123254365/230516403-5f8e388b-1b3d-402c-8515-fa9639edaa24.png)

Except when integrating it into the Energy dashboard, the solar generation energy entity it came up with this error. Trying to figure out now how to make the data an absolute value for Home Assistant.
![image](https://user-images.githubusercontent.com/123254365/230517084-5d52dee9-2c54-4791-bd34-a57233d1018b.png)

**Comment by isngofoz** [comment]
I figured out that to get the Solar Generation Sensor to read positive, I had to add 
value_template: "{{ value | multiply(-1) }}"  to the MQTT sensor.

Battery voltage for both the main sensor and solar sensor are reading now that the sensors are charged to full.

Probably more of my setup maybe, but I realised that it is not accurate to have the main sensor labelled Total Household Power as it actually measure the grid usage and grid export so the Total Household Power during the day is actually the solar generation plus grid usage if the main sensor is reading positive and solar generation minus grid export if the main sensor is reading negative. The Powersensor app represent it quite well.

**Comment by izevaka** [comment]
Very interesting. So the solar sensor reports the values as negative? That's kind of strange, but OK. As the luck will have it, my landlord wants to install solar, so I might play with a second powersensor plug in due course.

I think the net outcome here is that the node-red integration should take the absolute value for all power readings given that that's what Home Assistant is expecting.

Also, Im keen to get to the bottom of the sensor voltage reading. I can confirm that I am getting those readings and the sensor is not plugged in. Are you able to run some shell commands for me? Requries a mac or linux terminal to do so. 

Here is the command:

```
$> echo -n 'subscribe(30)\n' | nc -u IP_ADDRESS_PLUG 49476 | grep '"device":"sensor"'
```

Expected out:

```
{"mac":"bcddc247d630","device":"sensor","summation":785510302,"duration":30,"type":"instant_power","batteryMicrovolt":3860448,"unit":"w","starttime":1681887496,"power":544}
```

This should run for 30 seconds, but I interested in seeing both `"device":"sensor"` outputs - as in for the solar sensor and for the grid sensor.

**Comment by isngofoz** [comment]
I hope that I have done it right. Did it towards the end of the day though so solar production was very low.
![image](https://user-images.githubusercontent.com/123254365/233037837-66043fd3-ce4a-43b5-b6cb-d61838f663ae.png)

**Comment by izevaka** [comment]
Great thanks. That worked nicely. So... the sensor reports negative values (I assume that -21 W is the waning solar system). 

If we ever see the solar sensor going positive, we can probably tweak it a little bit to only report the absolute value it when it's negative like so.

```
{{ [value,0] | min | abs }}
```

Back to your suggestion about "Total Household Power" - totally agree, it needs a better label. Just brainstorming the different cases that can happen:

| Scenario | Main Sensor | Solar sensor |
|--------|--------|--------|
| Above zero | Household is importing from the grid | In theory shouldn't happen, unless the inverter is using the same wiring for grid sensing and self-power during the night  |
| Below zero | Household is exporting to the  grid| Solar system is producing electricity | 

OK, I have a follow-up question. Can you do another test run way after dark and see what your solar sensor is reading? Interested to see what happens after dark 😄

**Comment by isngofoz** [comment]
After dark result.
![image](https://user-images.githubusercontent.com/123254365/233082122-1b2e485f-2108-40d6-a095-c8859665bbc9.png)

Snapshot from configuration.yaml.
![image](https://user-images.githubusercontent.com/123254365/233082541-3e911037-02f8-4abe-8e82-ed8270fbb02c.png)

I am not familiar with yaml and search to web to find out how to get positive value and came up with the multiply(-1) solution. I will try out the one you have above and see what I get.

**Comment by isngofoz** [comment]
These are from the Energy dashboard today which is clearly confusing, as the end result is the grid owes me money which in actual fact if we have another grid export entity with a $ cost of feed-in, I will still owe money to the grid for using the energy. 

![image](https://user-images.githubusercontent.com/123254365/233086345-d44323d3-0a0b-4625-93c4-5be7887be3b8.png)
![image](https://user-images.githubusercontent.com/123254365/233086910-db0c3541-abc6-4795-a970-06e32e468f7c.png)
![image](https://user-images.githubusercontent.com/123254365/233087008-ee39eb37-3e12-4a22-a89a-44a8a1fce688.png)

Below from the Powersensor App.

![20230419_132230000_iOS](https://user-images.githubusercontent.com/123254365/233088486-fe669591-2404-4e58-afe3-8d978d726f7d.png)
![20230419_132242000_iOS](https://user-images.githubusercontent.com/123254365/233088537-729aa72e-9a68-41c7-b0e8-f202544c418d.png)

**Comment by izevaka** [comment]
Thanks for the additional screenshots. With regards to the YAML - don't stress too much. This change won't affect anything, it will just avoid double counting when the inverter is drawing from the grid. Which according to your latest screenshot it's not even doing :) 

The HA output looks strange, but in theory correct, it represents the net flow of electricity in and out of the house. I am expecting something that is more like this one (from https://community.home-assistant.io/t/monitor-your-solar-pv-system-in-home-assistant/342710):

![HA screenshot](https://community-assets.home-assistant.io/original/3X/1/6/1630e6a2f5c5d89a4dfb3b1f7eaaae760856e93f.png)

I suspect this is HA set up rather than the sensor output. Unfortunately I can't test end to end. But here is how I would set this up for the solar sensor:

![Screen Shot 2023-04-20 at 8 20 35 am](https://user-images.githubusercontent.com/384431/233212413-3a1ec293-397b-43a9-bb1f-f1a1d9fbbcf5.png)

**Comment by isngofoz** [comment]
I thought I will get you another set of data for when the solar system is operating in full sun.
![image](https://user-images.githubusercontent.com/123254365/233228634-f20b7d05-d68d-4ec7-9a52-810c92e37ff1.png)

**Comment by isngofoz** [comment]
Thanks for the link above. I will look into it a bit more when I get some time. 
I think the last time I concluded that Grid Export and Grid Import need to be an additional two separate entities, and some maths between the solar sensor and the sensor installed at the mains boards needs to be happen depending on the +ve and -ve values between the two sensors and the results stored in the two entities above every time the readings are polled/pushed. Or at least the Grid Export entity is required as HA allows to config a Grid Export, and that should probably even it out too. Anyways, too much YAML maths for me to try but I think ideally some maths code inside the node-red function block should fix the issue.

Either way, once you get your Solar sensor and Powersensor still has not released their HA integration, I am happy to assist with the trials.
![image](https://user-images.githubusercontent.com/123254365/233230154-64b6e2ae-d17b-4543-bb12-b0fe29a06be8.png)

**Comment by woopsicle** [comment]
I have had this working for a while now via code from this comment, with Solar. https://github.com/izevaka/powersensor-home-assistant/issues/1#issuecomment-1143180758 in case it helps?

**Comment by izevaka** [comment]
Thanks @woopsicle !!! Very helpful. Do you have a screenshot of your Energy dashboard config. I understand what you've done with positive/negative sensors now and it made me even more confused haha 😄

**Comment by woopsicle** [comment]
Sure, more screenshots attached. 

![Screenshot_20230420_101020_Home Assistant](https://user-images.githubusercontent.com/958704/233240085-f79b4b8f-d50b-42f0-a3da-b505e5d85e7c.jpg)
![Screenshot_20230420_101025_Home Assistant](https://user-images.githubusercontent.com/958704/233240099-f18f1644-f1d3-41d5-8d5e-8570eea4d7d8.jpg)
![Screenshot_20230420_101126_Home Assistant](https://user-images.githubusercontent.com/958704/233240113-40ed7ee7-3802-47f8-9f11-fa53f33b4de5.jpg)
![Screenshot_20230420_101136_Home Assistant](https://user-images.githubusercontent.com/958704/233240128-78a44d4a-84f5-445d-8c9a-66ff31bc876d.jpg)

**Comment by izevaka** [comment]
Awesome. Thanks. A bit trippy that HA can't figure out the direction of the power flow, but it makes sense now. Interesting that you used the utility meters. I didn't for my Energy dashboard as I found it was quite happy with the increasing counter.

@isngofoz Hope that's helpful. I'll try to test these out and update the YAML instructions in the readme, but it may take some creative sensor hacking :)

**Comment by isngofoz** [comment]
Thanks all. I will review what I have done and hopefully fill in the gaps correctly. Getting into HA, I was hoping more for a 'install integration' and 'woohoo' experience but I found that I am doing a fair bit more configuration than expected. All good either ways, just takes a bit more of my time to get to the 'woohoo' point.

**Comment by isngofoz** [comment]
Thanks guys. Managed to get it configured. It makes more sense now. Have not really compared it with the Powersensor App as yet. I did not use the Utility Meter as it was somewhat confusing to setup and since all I was after was a day to day comparison of energy which the energy dashboard appear to already support, I did not see a need.
![image](https://user-images.githubusercontent.com/123254365/234169322-ea5ba8fc-2a16-41b2-b428-3c61b2bb5c8a.png)

![image](https://user-images.githubusercontent.com/123254365/234169429-49490295-f372-488a-bbe0-e5581010942f.png)

Happy for this issue to be closed if there is nothing further to add.
